External/FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp

```diff
@@ -119,13 +188,107 @@ namespace FEXCore {
+         * passed to thunked APIs can safely be called by the native host library.
+         */
+        static void MakeHostTrampolineForGuestFunction(void* ArgsRV) {
+          struct ArgsRV_t {
+              IR::SHA256Sum *HostPackerSha256;
+              uintptr_t GuestUnpacker;
+              uintptr_t GuestTarget;
+              uintptr_t rv; // Pointer to host trampoline + TrampolineInstanceInfo
+          } *args = reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+          LOGMAN_THROW_A_FMT(args->GuestTarget, "Tried to create host-trampoline to null pointer guest function");
+
+          const auto CTX = Thread->CTX;
+          const auto ThunkHandler = reinterpret_cast<ThunkHandler_impl *>(CTX->ThunkHandler.get());
+
+          const GuestcallInfo gci = { args->GuestUnpacker, args->GuestTarget };
+
+          // Try first with shared_lock
+          {
+            std::shared_lock lk(ThunkHandler->ThunksMutex);
+
+            auto found = ThunkHandler->GuestcallToHostTrampoline.find(gci);
+            if (found != ThunkHandler->GuestcallToHostTrampoline.end()) {
+              args->rv = found->second;
+              return;
+            }
+          }
+
+          std::lock_guard lk(ThunkHandler->ThunksMutex);
+
+          // Retry lookup with full lock before making a new trampoline to avoid double trampolines
+          {
+            auto found = ThunkHandler->GuestcallToHostTrampoline.find(gci);
+            if (found != ThunkHandler->GuestcallToHostTrampoline.end()) {
+              args->rv = found->second;
+              return;
+            }
+          }
+
+          // No entry found => create new trampoline
+          auto HostPackerEntry = ThunkHandler->Thunks.find(*args->HostPackerSha256);
+          if (HostPackerEntry == ThunkHandler->Thunks.end()) {
+            ERROR_AND_DIE_FMT("Unknown host packing function for callback");
+          }
+
+          LogMan::Msg::DFmt("Thunks: Adding host trampoline for guest function {:#x}",
+                            args->GuestTarget);
+
+          const auto Length = __stop_HostToGuestTrampolineTemplate - __start_HostToGuestTrampolineTemplate;
+          const auto InstanceInfoOffset = Length - sizeof(TrampolineInstanceInfo);
+
+          if (ThunkHandler->HostTrampolineInstanceDataAvailable < Length) {
+            const auto allocation_step = 16 * 1024;
+            ThunkHandler->HostTrampolineInstanceDataAvailable = allocation_step;
+            ThunkHandler->HostTrampolineInstanceDataPtr = (uint8_t *)mmap(
+                0, ThunkHandler->HostTrampolineInstanceDataAvailable,
+                PROT_READ | PROT_WRITE | PROT_EXEC,
+                MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+
+            LOGMAN_THROW_A_FMT(ThunkHandler->HostTrampolineInstanceDataPtr != MAP_FAILED, "Failed to mmap HostTrampolineInstanceDataPtr");
+          }
+
+          const TrampolineInstanceInfo NewTrampolineInfo {
```

Comment:
Why not initialize directly in the target here, to avoid the memcpy?

Reply:
This is just being more explicit about creating a new object and overwriting any existing data in the trampoline section. Previously it could appear as if we're just patching a few specific items.

gcc and clang are predictable in optimizing away memcpy calls. There's some overhead if they can't infer that the target pointer is aligned, but considering the logic that leads up to this point I don't anticipate any practical performance difference.

Reply:
tbh, I find the initialize-and-copy way more confusing that just initializing inline myself - we can do like `*Dest = TrampolineInstanceInfo { ... }` which clearly initializes and copies, and we don't have to depend on fragile compiler optimisations :)